discovery/src/main/java/com/ibm/watson/developer_cloud/discovery/v1/Discovery.java

```diff
@@ -76,379 +45,803 @@
+   * the value of the `size` parameter. If you need to decrease an environment's size, contact IBM Support.
+   *
+   * @param environmentId the ID of your environment
+   * @param body
+   * @return the {@link Environment} with the response
+   */
+  public ServiceCall<Environment> updateEnvironment(String environmentId, UpdateEnvironmentRequest body) {
+    Validator.notEmpty(environmentId, "environmentId cannot be null");
+    Validator.notNull(body, "body cannot be null");
+    RequestBuilder builder = RequestBuilder.put(String.format("/v1/environments/%s", environmentId));
+    builder.query(VERSION, versionDate);
+    builder.bodyJson(GsonSingleton.getGson().toJsonTree(body).getAsJsonObject());
+
+    return createServiceCall(builder.build(), ResponseConverterUtils.getObject(Environment.class));
+  }
+
+  /**
+   * Query documents.
+   *
+   * See the [Query Language wiki page](https://github.ibm
+   * .com/Watson-Discovery/discovery-wiki/blob/master/design/query-language.md) for more details.
+   *
+   * @param environmentId the ID of your environment
+   * @param collectionId  the ID of your collection
+   * @param queryOptions  the query options
+   * @return the {@link QueryResponse} with the response
+   */
+  public ServiceCall<QueryResponse> query(String environmentId, String collectionId, QueryOptions queryOptions) {
+    Validator.notEmpty(environmentId, "environmentId cannot be null");
+    Validator.notEmpty(collectionId, "collectionId cannot be null");
+    RequestBuilder builder = RequestBuilder.get(String.format("/v1/environments/%s/collections/%s/query",
+        environmentId, collectionId));
+    builder.query(VERSION, versionDate);
+    if (queryOptions != null) {
+      if (queryOptions.filter() != null) {
+        builder.query("filter", RequestUtils.encode(queryOptions.filter()));
+      }
+      if (queryOptions.offset() != null) {
+        builder.query("offset", RequestUtils.encode(String.valueOf(queryOptions.offset())));
+      }
+      if (queryOptions.naturalLanguageQuery() != null) {
+        builder.query("natural_language_query", RequestUtils.encode(queryOptions.naturalLanguageQuery()));
+      }
+      if (queryOptions.query() != null) {
+        builder.query("query", RequestUtils.encode(queryOptions.query()));
+      }
+      if (queryOptions.count() != null) {
+        builder.query("count", RequestUtils.encode(String.valueOf(queryOptions.count())));
+      }
+      if (queryOptions.passages() != null) {
+        builder.query("passages", RequestUtils.encode(String.valueOf(queryOptions.passages())));
+      }
+      if (queryOptions.aggregation() != null) {
+        builder.query("aggregation", RequestUtils.encode(queryOptions.aggregation()));
+      }
+      if (queryOptions.sort() != null) {
+        builder.query("sort", RequestUtils.encode(queryOptions.sort()));
+      }
+      if (queryOptions.returnFields() != null) {
+        builder.query("return", queryOptions.returnFields());
```

Comment:
I agree that we should be encoding this array.  Unfortunately, I think the way to do this is to loop through the array encoding each element individually.  That's kinda ugly, but I suppose we do need to do it.

As for your point about doing encoding in the request handler, I agree that we should move in this direction.  If different services need different encodings, we need to write that code somewhere.  Doing it in the request handler will probably make that more efficient and maintainable.